pkg/apis/servicecatalog/unstructured_test.go

```diff
@@ -0,0 +1,152 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package servicecatalog_test
+
+import (
+	"math/rand"
+	"reflect"
+	"testing"
+
+	"github.com/google/gofuzz"
+
+	"github.com/kubernetes-incubator/service-catalog/pkg/apis/servicecatalog"
+	"github.com/kubernetes-incubator/service-catalog/pkg/apis/servicecatalog/testapi"
+	sctesting "github.com/kubernetes-incubator/service-catalog/pkg/apis/servicecatalog/testing"
+
+	"github.com/satori/go.uuid"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/conversion/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/diff"
+	"k8s.io/apimachinery/pkg/util/json"
+	"k8s.io/client-go/pkg/api"
+)
+
+// doUnstructuredRoundTrip performs the following round-tripping on a fuzzed
+// internal api object:
+//
+// 1. converts the internal object to versioned
+// 2. marshals the versioned object
+// 3. unmarshals the serialization of (2) into a map[string]interface{}
+// 4. marshals the unstructured object in (3)
+// 5. unmarshals (4) back into a versioned type
+// 6. does semantic equal on (5) and (1)
+// 7. uses the default converter to convert (1) directly into a map[string]interface{}
+// 8. converts (7) back into a versioned object
+// 9. does semantic equal on (1) and (8)
+func doUnstructuredRoundTrip(t *testing.T, group testapi.TestGroup, kind string) {
+	// We do fuzzing on the internal version of the object, and only then
+	// convert to the external version. This is because custom fuzzing
+	// function are only supported for internal objects.
+	internalObj, err := api.Scheme.New(group.InternalGroupVersion().WithKind(kind))
+	if err != nil {
+		t.Fatalf("Couldn't create internal object %v: %v", kind, err)
+	}
+	seed := rand.Int63()
```

Comment:
Would it be useful to stash away the used seed or log it for being able to repro test failures?

Reply:
We don't do that for other tests like this -- and I've never seen a failure in one of these tests that wasn't reproducible across runs that I can remember.

Reply:
Actually, there may be an issue here - seed for the rand is not initialized. Here is the godoc from the `rand.Seed()` function:
```
// Seed uses the provided seed value to initialize the default Source to a
// deterministic state. If Seed is not called, the generator behaves as
// if seeded by Seed(1). Seed values that have the same remainder when
// divided by 2^31-1 generate the same pseudo-random sequence.
// Seed, unlike the Rand.Seed method, is safe for concurrent use.
```
That means sequence is the same each run? Well, there may be some non-determinism introduced by something else, but this is not ideal.

This is how RNG is seeded in Smith for example (I don't see anything like that in svc-cat in the package with this test): https://github.com/atlassian/smith/blob/master/it/main_test.go